test/metabase/api/preview_embed_test.clj

```diff
@@ -141,6 +141,20 @@
              (mt/user-http-request :crowberto :get 202 (str (card-query-url card {:_embedding_params {:venue_id "enabled"}})
                                                             "?venue_id=200")))))))))
 
+(deftest query-max-results-constraint-test
+  (testing "GET /api/preview_embed/card/:token/query"
+    (testing "Only 2000 results returned when there are many more"
+      (embed-test/with-embedding-enabled-and-new-secret-key
+        (let [;; this is the same sample question as setup in these steps to reproduce:
+              ;; https://github.com/metabase/metabase/issues/20938
+              sample-db-orders-question {:database 1
+                                         :query {:source-table 2}
+                                         :type "query"}]
+          (embed-test/with-temp-card [card {:dataset_query sample-db-orders-question}]
+            (is (>= 2000 (-> (mt/user-http-request :crowberto :get 202 (card-query-url card))
+                             :data
+                             :rows
```

Comment:
Instead of `(-> ... :data :rows)` you should use `mt/rows` which does the same thing but throws a useful exception with the query error if the query failed. In that case, there is no `[:data :rows]` so `(count nil)` will just return `0` and your test failure will be 

```
(not (>= 2000 0))
```

which is not easy to debug. `mt/rows` will give you the actual query error in the failure

Reply:
```suggestion
                             mt/rows
```